src/main/java/dk/kb/util/yaml/YAML.java

```diff
@@ -668,52 +684,293 @@ public boolean containsKey(Object path) throws NullPointerException {
+        String key = conditionalMatch.group(1);
+        boolean mustMatch = conditionalMatch.group(2).equals("="); // The Pattern ensures only "!=" or "=" is in the group
+        String value = conditionalMatch.group(3);
+
+        List<Object> matchingObjects = new ArrayList<>();
+        for (Map.Entry<?,?> entry : map.entrySet()) {
+            matchingObjects.add(conditionalGet(entry.getValue(), key, value, mustMatch));
+        }
+
+        matchingObjects = matchingObjects.stream()
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+        if (yPath.size() <= 1){
+            for (Object matchingEntry: matchingObjects) {
+                visitor.visit(matchingEntry);
             }
-            if (sub == null) {
-                throw new NotFoundException(
-                        "Unable to request " + i + "'th sub-element: '" + pathKey + "'", path);
+        } else {
+            for (Object matchingEntry: matchingObjects) {
+                traverse(shortenedPath, matchingEntry, visitor);
             }
+        }
+    }
+
+
+    /**
+     * If the first element in a list starts with '[' and ends with ']' remove the brackets from the string. The brackets
+     * are used to describe a conditional lookup in a YAML map and needs to be removed before the traversal can be
+     * continued.
+     * @param yPath list of path elements, where the first element is the current element of the YAML traversal, which
+     *              is checked for '[' and ']'.
+     */
+    private static void removeBracketsFromPathElement(List<String> yPath) {
+        if (!yPath.isEmpty() && yPath.get(0).startsWith("[") && yPath.get(0).endsWith("]")){
+            String cleaned = yPath.get(0).substring(1, yPath.get(0).length()-1);
+            yPath.set(0, cleaned);
+        }
+    }
+
+    /**
+     * Convert a YAML list to a YAML map containing all elements, where the key is the index and continue the traversal
+     * of the YAML structure through the {@link #traverse(List, Object, MultipleValuesVisitor)}-method.
+     * @param yPath a list of path elements. This list contains all parts of a specified path, which is most likely
+     *              delivered through the {@link #visit(Object path, YAML yaml)}-method.
+     * @param visitor used to collect values that match the given path. The only function of this visitor is to collect
+     *               values to its internal list of extracted values: {@link MultipleValuesVisitor#extractedValues}.
+     * @param list the YAML list which is to be converted to a YAML map.
+     */
+    private void convertListToMapAndTraverse(List<String> yPath, MultipleValuesVisitor visitor, List<Object> list) {
+        LinkedHashMap<String, Object> map = new LinkedHashMap<>(list.size());
+        Object yaml;
+        for (int j = 0; j < list.size(); j++) {
+            map.put(j + "", extrapolate(list.get(j)));
+        }
+        yaml = map;
```

Comment:
This form of copying breaks path substitution. Consider the constructor
```Java
new YAML(map, extrapolateSystemProperties, getSubstitutors())
```
which transfers the structures needed.

I have added the unit test `testVisitPathSubstitutionSubmap` that triggers this.

Reply:
This will be solved by up front substitution.